skbio/stats/composition.py

```diff
@@ -0,0 +1,410 @@
+
+Many 'omics datasets are inheriently compositional - meaning that they are best
+interpreted as proportions or percentages rather than absolute counts.
+
+Formally, :math:`x` is a composition if :math:`\sum_{i=0}^D x_{i} = c` and
+:math:`x_{i} > 0`, :math:`1 \leq i \leq D`  and :math:`c` is a real valued
+constant and there are :math:`D` components for each composition. In this
+module :math:`c=1`. Compositional data can be analyzed using Aitchison
+geometry [1]_
+
+However, in this framework, standard real Euclidean operations such as addition
+and multiplication no longer apply.  Only operations such as perturbation and
+power can be used to manipulate this data [1]_
+
+This module allows two styles of manipulation of compositional data.
+Compositional data can be analyzed using perturbation and power operations,
+which can be useful for simulation studies. The alternative strategy is to
+transform compositional data into the real space.  Right now, the centre log
+ratio transform (clr) [1]_ can be used to accomplish this.  This transform can
+be useful for performing standard statistical tools such as parametric
+hypothesis testing, regressions and more.
+
+The major caveat of using this framework is dealing with zeros.
+In the Aitchison geometry, only compositions with nonzero components can be
+considered. The multiplicative replacement technique [2]_ can be used to
+substitute these zeros with small pseudocounts without introducing major
+distortions to the data.
+
+Functions
+---------
+
+.. autosummary::
+   :toctree: generated/
+
+   closure
+   multiplicative_replacement
+   perturb
+   perturb_inv
+   power
+   clr
+   centralize
+
+Examples
+--------
+
+>>> import numpy as np
+
+Consider a very simple environment with only 3 species.
+The species in the environment are equally distributed and their
+proportions are equilvalent:
+
+>>> otus = np.array([1./3, 1./3., 1./3])
+
+Suppose that an antibiotic kills off half of the population for the
+first two species, but doesn't harm the third species.
+Then the perturbation vector would be as follows
+
+>>> antibiotic = np.array([1./2, 1./2, 1])
+
+And the resulting perturbation would be
+
+>>> perturb(otus, antibiotic)
+array([ 0.25,  0.25,  0.5 ])
+
+Reference
+---------
+.. [1] V. Pawlowsky-Glahn. "Lecture Notes on Compositional Data Analysis"
+.. [2] J. A. Martin-Fernandez. "Dealing With Zeros and Missing Values in
+       Compositional Data Sets Using Nonparametric Imputation"
+
+"""
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from __future__ import absolute_import, division, print_function
+import numpy as np
+import scipy.stats as ss
+
+
+def closure(mat):
```

Comment:
That's what I thought when I first pushed this method.

However, when dealing with subcompositions, it is necessary to re-close the data in order to maintain that everything adds up to one.

For example, if there are 7 phyla you are interested in, but 3 of them are rare - you can exclude them and close the remaining 4 phyla so that their proportions add up to 1.  This a legitimate operation because it maintains [subcompositional coherence](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.324.1621&rep=rep1&type=pdf) (aka the ratios are preserved)

I'm already planning on using this technique when making tetrahedron plots for Emperor in the near future